src/packageurl/contrib/url2purl.py

```diff
@@ -276,3 +276,70 @@ def build_sourceforge_purl(uri):
 @purl_router.route(cargo_pattern)
 def build_cargo_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+#https://github.com/nexB/scancode-toolkit/tree/develop/plugins/scancode-ctags-macosx_10_9_intel
+@purl_router.route('https?://github.com/.*')
+def build_github_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
+
+    if len(segments) == 1 or len(segments) == 3:
+        return None 
+
+    if len(segments) == 2:
+        namespace = segments[0]
+        name = segments[1]
+        return PackageURL('github', name=name, namespace=namespace)
+    
+    if len(segments) >= 4:
+        if segments[2] == 'tree':
+            namespace = segments[0]
+            name = segments[1]
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+            return PackageURL('github', name=name, namespace=namespace, version=version, subpath=subpath)
+
+
+#https://bitbucket.org/TG1999/first_repo/src/master 
+@purl_router.route('https?://bitbucket.org/.*')
+def build_bitbucket_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
+
+    if len(segments) == 1 or (len(segments) >= 3 and segments[2] != 'src'):
+        return None 
+
+    if len(segments) == 2 or len(segments) == 3:
+        namespace = segments[0]
+        name = segments[1]
+        return PackageURL('bitbucket', name=name, namespace=namespace)
+    
+    if len(segments) >= 4:
+        namespace = segments[0]
+        name = segments[1]
+        version = segments[3]
+        subpath = '/'.join(segments[4:])
+        return PackageURL('bitbucket', name=name, namespace=namespace, version=version, subpath=subpath)
+
+
+#https://gitlab.com/TG1999/firebase/-/tree/1a122122/views
+@purl_router.route('https?://gitlab.com/.*')
+def build_gitlab_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
```

Comment:
Why repeat all this code in each of the github, bitbcket and gilab? you should have a function that does a base common parsing and returned some mini object or named tuple where each parts have a proper name.